lib/websocket_server.rb

```diff
@@ -1,56 +1,67 @@
+# This server allows the user to access remote endpoints through an HTTP(S)
+# connection to the appliance. Opposed to a direct access or a forwarded port
+# it is not necessary to expose an extra IP:port combination to the client.
+#
+# It uses socket hijacking to retrieve the underlying TCP socket connection
+# of the HTTP request incoming from a client. This socket is detached from
+# the Rack middleware and handled by a separate transmitter thread.
+# The remote address of the console is determined by a one time secret that
+# is part of the URL and it points to a SystemConsole record in the database.
+#
+# On the lowest level the transmitter thread operates with regular sockets,
+# however, both on the client and server side wrappers are used to translate
+# between a pair of sockets. Due to some limitations in the websocket driver
+# gem, the reading/writing operations are defined in the fetch/issue methods.
+#
+# The selector for the sockets is provided by an external gem and it handles
+# the dependency between the socket pairs, i.e. one should be ready to read
+# while other one should be ready to write at the same time. The pairs ready
+# for transmission are handled by the `each_ready` iterator. As the iterator
+# always returns with a socket to read and a socket to write, the `@adapters`
+# hash has been used to access the corresponding wrappers.
+
 class WebsocketServer
   attr_accessor :logger
 
-  Pairing = Struct.new(:is_ws, :proxy)
+  RACK_404 = [404, {'Content-Type' => 'text/plain'}, ['Not found']].freeze
+  RACK_YAY = [-1, {}, []].freeze
 
   def initialize(options = {})
     @logger = options.fetch(:logger, $websocket_log)
-    logger.info('Initializing websocket worker!')
-    @pairing = {}
-    @sockets = Concurrent::Array.new
+    @logger.info('Initializing websocket server!')
+    @proxy = SurroGate.new(logger)
+    @adapters = {}
 
     @transmitter = Thread.new do
       loop do
-        begin
-          reads, writes, errors = IO.select(@sockets, @sockets, @sockets, 1)
-        rescue IOError
-          @sockets.select(&:closed?).each { |err| cleanup(err) }
-        else
-          Array(errors).each { |err| cleanup(err) }
-        end
-
-        # Skip this loop if we can't do anything
-        if Array(reads).empty? || writes.empty? || errors.any?
-          sleep(1) if @sockets.empty?
-          next
-        end
+        @proxy.select(1000)
 
-        # Do the data transfers
-        reads.each do |socket|
+        @proxy.each_ready do |left, right|
           begin
-            @pairing[socket].proxy.transmit(writes, @pairing[socket].is_ws)
-          rescue => error
-            cleanup(socket, error)
+            @adapters[left].fetch(64.kilobytes) { |data| @adapters[right].issue(data) } # left -> right
+          rescue IOError, EOFError
```

Comment:
I think the "Do not shadow rescued Exceptions" cop wants these exceptions swapped so the subclass "EOFERROR" is before the ancestor class "IOERROR"

Reply:
Oh, or I can just drop the second one as it is a subclass ... :1st_place_medal:

Reply:
haha, yup... I was so confused why rubocop was complaining that I missed the most obvious thing.  🔴 🚽 🗑